File: evaluation/support-program-search/review/test_review_page.py
import copy
import importlib.util
import json
import tempfile
import unittest
from pathlib import Path
from types import SimpleNamespace
from unittest.mock import patch


REVIEW_DIR = Path(__file__).resolve().parent


def load_module(name, filename):
    spec = importlib.util.spec_from_file_location(name, REVIEW_DIR / filename)
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    return module


PAGE = load_module("review_page_test_subject", "build-review-page.py")
EXTRACT = load_module("review_json_test_subject", "extract-review-json.py")
SUPPORT = load_module("existing_review_test_scenarios", "test_review_tools.py")


class BrowserReviewTest(unittest.TestCase):
    def setUp(self):
        self.temp = tempfile.TemporaryDirectory()
        self.root = Path(self.temp.name)
        self.scenario = SUPPORT.create_scenario(self.root)
        self.pool_args = SUPPORT.build_pool(self.scenario)
        self.fixture = PAGE.load_json(self.scenario.paths["fixture"])
        self.query_set = PAGE.load_json(self.scenario.paths["query_set"])
        self.manifest = PAGE.load_json(self.pool_args.pool_manifest)
        self.rows = SUPPORT.read_rows(self.pool_args.review_pool)
        self.template_dir = self.root / "templates"
        self.template_dir.mkdir()
        (self.template_dir / "review-page.html").write_text("<script type=\"application/json\" id=\"data\">__REVIEW_DATA__</script>", encoding="utf-8")

    def tearDown(self):
        self.temp.cleanup()

    def seed_file(self):
        seeds = {
            "referenceDate": self.fixture["referenceDate"],
            "querySetSha256": self.manifest["querySetSha256"],
            "reviewer": "대화 사용자",
            "reviewMethod": "대화에서 제시된 요약을 읽고 판단함",
            "judgments": [
                {
                    "queryId": "Q01", "programId": self.fixture["docs"][index]["id"],
                    "contentHash": self.fixture["docs"][index]["contentHash"],
                    "presentedQuery": "알파 지원을 찾고 있어",
                    "presentedProgramTitle": f"알파 지원 {index + 1}",
                    "presentedProgramSummary": "대화에서 보여준 짧은 요약",
                    "decision": "relevant" if index == 0 else "irrelevant",
                    "userResponse": "추천해도 되지" if index == 0 else "안 되지 지역이 다르니까",
                    "userReason": None if index == 0 else "지역이 다르니까",
                }
                for index in range(2)
            ],
        }
        path = self.root / "conversation.json"
        SUPPORT.write_json(path, seeds)
        return path, seeds

    def page_args(self, seeds=None, name="review.html"):
        return SimpleNamespace(
            fixture=self.scenario.paths["fixture"], query_set=self.scenario.paths["query_set"],
            review_pool=self.pool_args.review_pool, pool_manifest=self.pool_args.pool_manifest,
            conversation_judgments=str(seeds) if seeds else None, output=str(self.root / name),
        )

    def build_page(self, args):
        with patch.object(PAGE, "REVIEW_DIR", self.template_dir):
            return SUPPORT.call_main(PAGE, args)

    def export(self):
        return {
            **PAGE.review_identity(self.manifest), "reviewer": "검토자",
            "judgments": [
                {"queryId": row["query_id"], "programId": row["program_id"],
                 "decision": "", "reason": "", "reviewer": "", "provenance": None}
                for row in self.rows
            ],
        }

    def test_build_preserves_two_human_judgments_and_raw_conversation(self):
        seeds_path, seeds = self.seed_file()
        args = self.page_args(seeds_path)
        self.build_page(args)
        html = Path(args.output).read_text(encoding="utf-8")
        payload = json.loads(html.split(">", 1)[1].rsplit("</script>", 1)[0])
        self.assertEqual(payload["rows"], self.rows)
        self.assertEqual(len(payload["seedJudgments"]), 2)
        first, second = payload["seedJudgments"]
        self.assertEqual(first["decision"], "relevant")
        self.assertIn("별도 사유 미입력", first["reason"])
        self.assertEqual(first["provenance"]["userResponse"], "추천해도 되지")
        self.assertIsNone(first["provenance"]["userReason"])
        self.assertEqual(first["provenance"]["presentedProgramSummary"], seeds["judgments"][0]["presentedProgramSummary"])
        self.assertEqual(second["decision"], "irrelevant")
        self.assertEqual(second["reason"], "지역이 다르니까")
        self.assertEqual(first["reviewer"], "대화 사용자")

    def test_build_refuses_tampered_fixture_hash_and_catalog(self):
        for kind in ("content", "fingerprint", "count"):
            with self.subTest(kind=kind):
                fixture = copy.deepcopy(self.fixture)
                if kind == "content":
                    fixture["docs"][0]["text"] += "tampered"
                elif kind == "fingerprint":
                    fixture["catalog"]["eligibleCatalogFingerprint"] = "0" * 64
                else:
                    fixture["catalog"]["eligibleProgramCount"] -= 1
                SUPPORT.write_json(Path(self.scenario.paths["fixture"]), fixture)
                with self.assertRaises(ValueError):
                    self.build_page(self.page_args())
                self.assertFalse((self.root / "review.html").exists())

    def test_build_refuses_changed_query_and_manifest_identity(self):
        query_set = copy.deepcopy(self.query_set)
        query_set["queries"][0]["query"] = "다른 질문"
        with self.assertRaisesRegex(ValueError, "query-set hash"):
            PAGE.validate_sources(self.fixture, query_set, self.manifest, self.rows)
        for field, value in (("name", "other"), ("referenceDate", "2026-09-05"), ("catalogFingerprint", "0" * 64)):
            with self.subTest(field=field):
                manifest = {**self.manifest, field: value}
                with self.assertRaises(ValueError):
                    PAGE.validate_sources(self.fixture, self.query_set, manifest, self.rows)

    def test_refuses_immutable_tamper_even_if_manifest_structure_was_changed(self):
        rows = copy.deepcopy(self.rows)
        rows[0]["summary"] = "거짓 요약"
        manifest = {**self.manifest, "reviewStructureSha256": PAGE.POOL.review_structure_sha256(rows)}
        with self.assertRaisesRegex(ValueError, "fixture text"):
            PAGE.validate_sources(self.fixture, self.query_set, manifest, rows)

    def test_pool_rejects_missing_duplicate_unknown_or_changed_rows(self):
        for kind in ("missing", "duplicate", "unknown", "changed", "bad-decision"):
            with self.subTest(kind=kind):
                rows = copy.deepcopy(self.rows)
                if kind == "missing":
                    rows.pop()
                elif kind == "duplicate":
                    rows[-1] = rows[0]
                elif kind == "unknown":
                    rows[0]["query_id"] = "UNKNOWN"
                elif kind == "changed":
                    rows[0]["title"] += "변경"
                else:
                    rows[0]["decision"] = "maybe"
                path = self.root / f"{kind}.csv"
                SUPPORT.write_rows(path, rows)
                with self.assertRaises(ValueError):
                    PAGE.load_verified_pool(path, self.manifest)

    def test_seed_rejects_wrong_source_identity_unknown_duplicate_or_conflict(self):
        path, seeds = self.seed_file()
        for kind in ("hash", "query-hash", "date", "unknown", "duplicate", "blank", "conflict"):
            with self.subTest(kind=kind):
                value = copy.deepcopy(seeds)
                rows = copy.deepcopy(self.rows)
                if kind == "hash":
                    value["judgments"][0]["contentHash"] = "0" * 64
                elif kind == "query-hash":
                    value["querySetSha256"] = "0" * 64
                elif kind == "date":
                    value["referenceDate"] = "2026-09-05"
                elif kind == "unknown":
                    value["judgments"][0]["programId"] = "SYNTHETIC:UNKNOWN"
                elif kind == "duplicate":
                    value["judgments"].append(value["judgments"][0])
                elif kind == "blank":
                    value["judgments"][0]["decision"] = ""
                else:
                    row = next(row for row in rows if (row["query_id"], row["program_id"]) == ("Q01", "SYNTHETIC:1"))
                    row["decision"] = "irrelevant"
                SUPPORT.write_json(path, value)
                with self.assertRaises(ValueError):
                    PAGE.load_seeds(path, self.fixture, self.manifest, rows)

    def test_inline_json_escapes_script_breakout_and_unicode_separators(self):
        raw = {"text": '</script><img src=x onerror="alert(1)"> & \u2028 \u2029'}
        encoded = PAGE.script_json(raw)
        self.assertEqual(json.loads(encoded), raw)
        for forbidden in ("<", ">", "&", "\u2028", "\u2029"):
            self.assertNotIn(forbidden, encoded)

    def test_seed_export_roundtrip_retains_decisions_without_fabricating_other_labels(self):
        path, seeds = self.seed_file()
        seed_judgments = PAGE.load_seeds(path, self.fixture, self.manifest, self.rows)
        seeded = {(item["queryId"], item["programId"]): item for item in seed_judgments}
        exported = self.export()
        exported["judgments"] = [
            seeded.get((item["queryId"], item["programId"]), item)
            for item in exported["judgments"]
        ]
        serialized = PAGE.script_json(exported)
        restored = json.loads(serialized)
        converted = EXTRACT.apply_judgments(self.rows, self.manifest, restored)
        self.assertEqual(sum(PAGE.is_complete(row) for row in converted), 2)
        self.assertEqual(sum(not row["decision"] for row in converted), len(self.rows) - 2)
        for judgment in restored["judgments"]:
            if judgment["provenance"]:
                original = next(item for item in seeds["judgments"] if item["programId"] == judgment["programId"])
                self.assertEqual(judgment["provenance"]["userResponse"], original["userResponse"])
                self.assertEqual(judgment["provenance"]["userReason"], original["userReason"])

    def test_existing_csv_draft_is_accepted_and_not_marked_complete(self):
        rows = copy.deepcopy(self.rows)
        rows[0].update(decision="relevant", reason="", reviewer="검토자")
        path = self.root / "draft.csv"
        SUPPORT.write_rows(path, rows)
        restored = PAGE.load_verified_pool(path, self.manifest)
        self.assertEqual(restored[0]["decision"], "relevant")
        self.assertFalse(PAGE.is_complete(restored[0]))

    def test_build_refuses_existing_output_without_changing_it(self):
        args = self.page_args()
        Path(args.output).write_text("기존 검토 결과", encoding="utf-8")
        with self.assertRaises(FileExistsError):
            self.build_page(args)
        self.assertEqual(Path(args.output).read_text(encoding="utf-8"), "기존 검토 결과")

    def test_build_requires_one_template_token(self):
        (self.template_dir / "review-page.html").write_text("__REVIEW_DATA____REVIEW_DATA__", encoding="utf-8")
        with self.assertRaisesRegex(ValueError, "exactly one"):
            self.build_page(self.page_args())
        self.assertFalse((self.root / "review.html").exists())

    def test_json_loader_rejects_duplicate_fields(self):
        path = self.root / "duplicate.json"
        path.write_text('{"decision":"relevant","decision":"irrelevant"}', encoding="utf-8")
        with self.assertRaisesRegex(ValueError, "Duplicate JSON field"):
            PAGE.load_json(path)

    def test_convert_preserves_partial_progress_and_all_original_fields(self):
        exported = self.export()
        exported["judgments"][0].update(decision="relevant", reason="작성 중인 사유", reviewer="")
        exported["judgments"][1].update(decision="unclear", reason="", reviewer="작성자")
        exported["judgments"][2].update(decision="irrelevant", reason="지역 불일치", reviewer="작성자")
        exported["judgments"][3].update(reason="선택 전 메모", reviewer="작성자")
        exported["judgments"].reverse()
        path = self.root / "browser.json"
        SUPPORT.write_json(path, exported)
        args = SimpleNamespace(review_pool=self.pool_args.review_pool, pool_manifest=self.pool_args.pool_manifest,
                               review_json=str(path), output=str(self.root / "converted.csv"))
        stdout = SUPPORT.call_main(EXTRACT, args)
        report = json.loads(stdout)
        self.assertEqual(report["completedRowCount"], 1)
        self.assertEqual(report["incompleteRowCount"], 2)
        self.assertEqual(report["unselectedRowCount"], 1)
        converted = SUPPORT.read_rows(args.output)
        self.assertEqual(PAGE.POOL.review_structure_sha256(converted), self.manifest["reviewStructureSha256"])
        self.assertEqual([(row["query_id"], row["program_id"]) for row in converted], [(row["query_id"], row["program_id"]) for row in self.rows])
        self.assertEqual(converted[0]["reason"], "작성 중인 사유")
        self.assertEqual(converted[1]["decision"], "unclear")
        self.assertEqual(converted[3]["reason"], "선택 전 메모")
        self.assertEqual(converted[3]["decision"], "")

    def test_convert_exact_identity_and_row_contract(self):
        for kind in ("hash", "capture", "missing", "duplicate", "unknown", "extra", "missing-field", "immutable", "bad-decision", "bad-reason", "bad-reviewer", "bad-provenance"):
            with self.subTest(kind=kind):
                exported = self.export()
                if kind == "hash":
                    exported["reviewStructureSha256"] = "0" * 64
                elif kind == "capture":
                    exported["captureIncluded"] = 1
                elif kind == "missing":
                    exported["judgments"].pop()
                elif kind == "duplicate":
                    exported["judgments"][-1] = exported["judgments"][0]
                elif kind == "unknown":
                    exported["judgments"][0]["programId"] = "SYNTHETIC:UNKNOWN"
                elif kind == "extra":
                    exported["unexpected"] = True
                elif kind == "missing-field":
                    del exported["judgments"][0]["provenance"]
                elif kind == "immutable":
                    exported["judgments"][0]["title"] = "외부 제목"
                elif kind == "bad-decision":
                    exported["judgments"][0]["decision"] = "yes"
                elif kind == "bad-reason":
                    exported["judgments"][0]["reason"] = "가" * 2001
                elif kind == "bad-reviewer":
                    exported["judgments"][0]["reviewer"] = "가" * 101
                else:
                    exported["judgments"][0]["provenance"] = "not an object"
                with self.assertRaises(ValueError):
                    EXTRACT.apply_judgments(self.rows, self.manifest, exported)

    def test_convert_refuses_existing_output_and_original_csv(self):
        path = self.root / "browser.json"
        SUPPORT.write_json(path, self.export())
        original = Path(self.pool_args.review_pool).read_bytes()
        args = SimpleNamespace(review_pool=self.pool_args.review_pool, pool_manifest=self.pool_args.pool_manifest,
                               review_json=str(path), output=self.pool_args.review_pool)
        with self.assertRaises(FileExistsError):
            SUPPORT.call_main(EXTRACT, args)
        self.assertEqual(Path(self.pool_args.review_pool).read_bytes(), original)


if __name__ == "__main__":
    unittest.main()

File: evaluation/support-program-search/review/test-review-page.mjs
// Run with the bundled Node runtime and explicit REVIEW_PLAYWRIGHT_MODULE.
// REVIEW_BROWSER_EXECUTABLE may point to an already installed Chrome binary.
// No packages/browsers are installed and only temporary synthetic data is used.
import assert from "node:assert/strict";
import { spawnSync } from "node:child_process";
import { createHash } from "node:crypto";
import fs from "node:fs/promises";
import http from "node:http";
import os from "node:os";
import path from "node:path";
import { fileURLToPath, pathToFileURL } from "node:url";
import test from "node:test";

const directory = path.dirname(fileURLToPath(import.meta.url));
const modulePath = process.env.REVIEW_PLAYWRIGHT_MODULE;
const python = process.env.REVIEW_PYTHON || "python3";
const headers = [
  "query_id", "split", "query", "decision", "reason", "reviewer", "title",
  "summary", "target_description", "regions", "categories", "application_period",
  "organization", "program_id",
];
const mutable = new Set(["decision", "reason", "reviewer"]);
const sha256 = (text) => createHash("sha256").update(text).digest("hex");
const compact = (items) => JSON.stringify(items.map((item) => Object.fromEntries(
  Object.entries(item).sort(([left], [right]) => left < right ? -1 : left > right ? 1 : 0),
)));
const csv = (rows) => "\uFEFF" + [headers, ...rows.map((row) => headers.map((field) => row[field]))]
  .map((row) => row.map((value) => /[",\r\n]/.test(value) ? '"' + value.replaceAll('"', '""') + '"' : value).join(","))
  .join("\r\n") + "\r\n";

async function setup(temporary) {
  const rows = ["001", "002", "003"].map((id, index) => ({
    query_id: "Q01", split: "dev", query: "서울 소상공인의 AI 홍보 콘텐츠 제작 지원",
    decision: "", reason: "", reviewer: "",
    title: ["서울 AI 콘텐츠 제작", "울산 마케팅 지원", "아직 확인하지 않은 합성 공고"][index],
    summary: '한글 "인용"과 줄바꿈\n' + "공고에 없는 조건은 추측하지 않습니다. ".repeat(10)
      + '<img src="https://example.invalid/leak" onerror="window.injected=true">',
    target_description: "소상공인", regions: index === 1 ? "울산" : "서울",
    categories: "AI", application_period: "상시", organization: "합성 기관",
    program_id: "SYNTHETIC:" + id,
  }));
  const manifest = {
    schemaVersion: "support-program-review-pool-manifest-v1",
    name: "synthetic-browser-test", referenceDate: "2026-09-06",
    querySetSha256: sha256("synthetic query set"), captureIncluded: false,
    reviewRowCount: rows.length, perQueryCounts: { Q01: rows.length },
    reviewStructureSha256: sha256(compact(rows.map((row) => Object.fromEntries(
      Object.entries(row).filter(([field]) => !mutable.has(field)),
    )))),
    poolKeySha256: sha256(rows.map((row) => row.query_id + "\t" + row.program_id).sort().join("\n")),
  };
  const seeds = rows.slice(0, 2).map((row, index) => ({
    queryId: row.query_id, programId: row.program_id,
    decision: index === 0 ? "relevant" : "irrelevant",
    reason: index === 0 ? "대화에서 요약을 읽고 추천 가능 판정(별도 사유 미입력)" : "서울 질문인데 울산 대상",
    reviewer: "대화 사용자",
    provenance: {
      kind: "conversation", basis: "conversation_summary",
      userResponse: index === 0 ? "추천해도 되지" : "안되지 대상이 질문자는 서울인데, 대상자는 울산이니까",
      userReason: index === 0 ? null : "지역이 다름",
      presentedQuery: rows[0].query, presentedProgramTitle: row.title,
      presentedProgramSummary: "짧은 합성 요약", reviewMethod: "대화 요약 확인",
    },
  }));
  const payload = {
    schemaVersion: "support-program-browser-review-v1", name: manifest.name,
    referenceDate: manifest.referenceDate, querySetSha256: manifest.querySetSha256,
    reviewStructureSha256: manifest.reviewStructureSha256,
    poolKeySha256: manifest.poolKeySha256, captureIncluded: manifest.captureIncluded,
    rows, seedJudgments: seeds,
  };
  const encoded = JSON.stringify(payload).replaceAll("<", "\\u003c").replaceAll(">", "\\u003e");
  const template = await fs.readFile(path.join(directory, "review-page.html"), "utf8");
  assert.equal(template.split("__REVIEW_DATA__").length, 2);
  const html = template.replace("__REVIEW_DATA__", encoded);
  await fs.writeFile(path.join(temporary, "review.html"), html);
  await fs.writeFile(path.join(temporary, "review-pool.csv"), csv(rows));
  await fs.writeFile(path.join(temporary, "review-pool-manifest.json"), JSON.stringify(manifest));
  return { rows, seeds, manifest, html };
}

test("offline review page: isolated synthetic browser and converter workflow", {
  skip: !modulePath ? "Set REVIEW_PLAYWRIGHT_MODULE to the bundled Playwright module; nothing is installed automatically." : false,
  timeout: 120000,
}, async (t) => {
  const imported = await import(pathToFileURL(path.resolve(modulePath)).href);
  const { chromium } = imported.default || imported;
  const executablePath = process.env.REVIEW_BROWSER_EXECUTABLE || chromium.executablePath();
  await fs.access(executablePath);
  const temporary = await fs.mkdtemp(path.join(os.tmpdir(), "govbiz-review-browser-test-"));
  let browser;
  let server;
  const externalRequests = [];
  const pageErrors = [];
  try {
    const scenario = await setup(temporary);
    server = http.createServer((request, response) => {
      if (request.url !== "/review.html") {
        response.writeHead(404);
        response.end();
        return;
      }
      response.writeHead(200, { "content-type": "text/html; charset=utf-8" });
      response.end(scenario.html);
    });
    await new Promise((resolve, reject) => {
      server.once("error", reject);
      server.listen(0, "127.0.0.1", resolve);
    });
    const origin = "http://127.0.0.1:" + server.address().port;
    const url = origin + "/review.html";
    browser = await chromium.launch({ executablePath, headless: true, args: [
      "--disable-background-networking", "--disable-component-update", "--disable-sync",
      "--no-first-run", "--no-default-browser-check",
    ] });
    async function isolatedContext(options = {}) {
      const context = await browser.newContext({ acceptDownloads: true, ...options });
      await context.route("**/*", async (route) => {
        if (route.request().url() === url) await route.continue();
        else {
          externalRequests.push(route.request().url());
          await route.abort();
        }
      });
      context.on("page", (page) => page.on("pageerror", (error) => pageErrors.push(error.message)));
      return context;
    }
    const context = await isolatedContext({ viewport: { width: 1100, height: 900 } });
    const page = await context.newPage();
    await page.goto(url);
    const count = () => page.locator("#progress-text").textContent();
    let exportNumber = 0;
    async function download(pageToUse = page) {
      const waiting = pageToUse.waitForEvent("download");
      await pageToUse.locator("#export-button").click();
      const file = await waiting;
      const output = path.join(temporary, "export-" + (++exportNumber) + ".json");
      await file.saveAs(output);
      return { output, data: JSON.parse(await fs.readFile(output, "utf8")) };
    }
    async function importValue(value, accept = false) {
      await page.locator("#message").evaluate((element) => { element.textContent = ""; });
      if (accept) page.once("dialog", (dialog) => dialog.accept());
      await page.locator("#import-file").setInputFiles({
        name: "review-progress.json", mimeType: "application/json",
        buffer: Buffer.from(JSON.stringify(value)),
      });
      await page.locator("#message").filter({ hasText: accept ? "불러왔습니다" : "불러오지 않았습니다" }).waitFor();
    }

    await t.test("two conversation judgments are counted and source text is not executed", async () => {
      assert.match(await count(), /2 \/ 3건 입력 완료/);
      assert.match(await page.locator("#program-title").textContent(), /아직 확인하지 않은/);
      assert.equal(await page.locator("#program-summary img").count(), 0);
      assert.equal(await page.evaluate(() => window.injected), undefined);
      const initial = await download();
      assert.equal(initial.data.judgments.length, 3);
      assert.deepEqual(initial.data.judgments.slice(0, 2), scenario.seeds);
    });
    await t.test("a selected recommendation needs a reason and reviewer to be complete", async () => {
      await page.locator('[data-decision="relevant"]').click();
      assert.match(await count(), /2 \/ 3/);
      assert.match(await page.locator("#current-status").textContent(), /이유/);
      await page.locator("#reason").fill("질문의 지역과 지원 목적이 일치합니다.");
      assert.match(await count(), /3 \/ 3/);
      await page.locator("#reviewer").fill("");
      assert.match(await count(), /2 \/ 3/);
      await page.locator("#reviewer").fill("합성 검토자");
      assert.match(await count(), /3 \/ 3/);
    });
    await t.test("navigation preserves judgments and displays the original conversation", async () => {
      await page.locator("#previous").click();
      assert.match(await page.locator("#seed-note").textContent(), /질문자는 서울/);
      await page.locator("#previous").click();
      assert.match(await page.locator("#seed-note").textContent(), /추천해도 되지/);
      assert.equal(await page.locator("#previous").isDisabled(), true);
      await page.locator("#next").click();
      await page.locator("#next").click();
      assert.equal(await page.locator("#reason").inputValue(), "질문의 지역과 지원 목적이 일치합니다.");
      await page.locator("#next-unfinished").click();
      assert.match(await page.locator("#message").textContent(), /입력이 끝났습니다/);
    });
    await t.test("refresh restores both progress and position from local storage", async () => {
      await page.reload();
      assert.match(await count(), /3 \/ 3/);
      assert.match(await page.locator("#program-title").textContent(), /아직 확인하지 않은/);
      assert.equal(await page.locator("#reason").inputValue(), "질문의 지역과 지원 목적이 일치합니다.");
      assert.equal(await page.locator("#reviewer").inputValue(), "합성 검토자");
    });
    await t.test("unclear requires a reason and remains separately visible", async () => {
      await page.locator('[data-decision="unclear"]').click();
      await page.locator("#reason").fill("");
      assert.match(await count(), /2 \/ 3/);
      await page.locator("#reason").fill("핵심 조건이 요약에 없습니다.");
      assert.match(await count(), /3 \/ 3/);
      assert.match(await page.locator("#counts").textContent(), /판단 어려움 1건/);
    });
    let draft;
    await t.test("export retains all rows, the two seeds, and an unfinished draft", async () => {
      page.once("dialog", (dialog) => dialog.accept());
      await page.locator("#clear").click();
      await page.locator("#reason").fill('아직 작성 중인 이유\n한글 "인용"');
      draft = await download();
      assert.equal(draft.data.judgments.length, 3);
      assert.deepEqual(draft.data.judgments.slice(0, 2), scenario.seeds);
      assert.equal(draft.data.judgments[2].decision, "");
      assert.equal(draft.data.judgments[2].reason, '아직 작성 중인 이유\n한글 "인용"');
      assert.match(await count(), /2 \/ 3/);
    });
    await t.test("invalid imports are rejected atomically without changing current progress", async () => {
      const bad = [
        { ...draft.data, querySetSha256: "0".repeat(64) },
        { ...draft.data, judgments: draft.data.judgments.slice(0, 2) },
        { ...draft.data, judgments: [...draft.data.judgments.slice(0, 2), draft.data.judgments[0]] },
        { ...draft.data, judgments: draft.data.judgments.map((item, index) => index === 2 ? { ...item, decision: "maybe" } : item) },
      ];
      for (const value of bad) {
        await importValue(value);
        assert.deepEqual((await download()).data, draft.data);
      }
    });
    await t.test("explicit valid import restores unfinished drafts and original provenance", async () => {
      await page.locator('[data-decision="irrelevant"]').click();
      await page.locator("#reason").fill("다른 임시 입력");
      await importValue(draft.data, true);
      assert.deepEqual((await download()).data, draft.data);
      assert.equal(await page.locator("#reason").inputValue(), draft.data.judgments[2].reason);
      assert.match(await count(), /2 \/ 3/);
    });
    await t.test("downloaded JSON converts to the full matching CSV and cannot overwrite outputs", async () => {
      const output = path.join(temporary, "converted.csv");
      const args = ["-B", path.join(directory, "extract-review-json.py"),
        "--review-pool", path.join(temporary, "review-pool.csv"),
        "--pool-manifest", path.join(temporary, "review-pool-manifest.json"),
        "--review-json", draft.output, "--output", output];
      const result = spawnSync(python, args, { encoding: "utf8", timeout: 15000 });
      assert.equal(result.status, 0, result.stderr);
      const expected = scenario.rows.map((row, index) => ({
        ...row, decision: draft.data.judgments[index].decision,
        reason: draft.data.judgments[index].reason, reviewer: draft.data.judgments[index].reviewer,
      }));
      assert.equal(await fs.readFile(output, "utf8"), csv(expected));
      assert.equal(JSON.parse(result.stdout).reviewRowCount, 3);
      const second = spawnSync(python, args, { encoding: "utf8", timeout: 15000 });
      assert.notEqual(second.status, 0);
      assert.match(second.stderr, /already exists/);
      assert.equal(await fs.readFile(output, "utf8"), csv(expected));
    });
    await t.test("blocked local storage warns but still supports a complete downloadable backup", async () => {
      const unavailable = await isolatedContext();
      await unavailable.addInitScript(() => {
        for (const method of ["getItem", "setItem"]) {
          Storage.prototype[method] = () => { throw new DOMException("Disabled for test", "SecurityError"); };
        }
      });
      const blockedPage = await unavailable.newPage();
      await blockedPage.goto(url);
      assert.equal(await blockedPage.locator("#warning").isVisible(), true);
      assert.match(await blockedPage.locator("#storage-status").textContent(), /사용 불가|저장 필요/);
      const backup = await download(blockedPage);
      assert.equal(backup.data.judgments.length, 3);
      assert.deepEqual(backup.data.judgments.slice(0, 2), scenario.seeds);
      await unavailable.close();
    });
    await t.test("a second window cannot silently overwrite another window's changes", async () => {
      const secondPage = await context.newPage();
      await secondPage.goto(url);
      await secondPage.locator("#reason").fill("두 번째 창 입력");
      await page.locator("#warning").waitFor();
      assert.match(await page.locator("#warning").textContent(), /다른 창/);
      await page.locator("#reason").fill("첫 번째 창의 보관할 입력");
      await secondPage.reload();
      assert.equal(await secondPage.locator("#reason").inputValue(), "두 번째 창 입력");
      await secondPage.close();
    });
    await t.test("360px mobile layout has no horizontal overflow and all decision buttons are usable", async () => {
      const mobile = await isolatedContext({ viewport: { width: 360, height: 800 } });
      const mobilePage = await mobile.newPage();
      await mobilePage.goto(url);
      assert.equal(await mobilePage.evaluate(() => document.documentElement.scrollWidth <= innerWidth), true);
      for (const decision of ["relevant", "irrelevant", "unclear"]) {
        const button = mobilePage.locator('[data-decision="' + decision + '"]');
        await button.click();
        assert.equal(await button.getAttribute("aria-pressed"), "true");
        assert.equal(await mobilePage.evaluate(() => document.documentElement.scrollWidth <= innerWidth), true);
      }
      await mobile.close();
    });
    await t.test("hybrid mode requires selected human checks without prefilling AI as human", async () => {
      const generated = spawnSync(python, ["-B", "-c", `
import importlib.util, json, sys
spec = importlib.util.spec_from_file_location("mode_browser_synthetic", sys.argv[1])
module = importlib.util.module_from_spec(spec)
spec.loader.exec_module(module)
scenario = module.ReviewModeTest()
scenario.setUp()
try:
    selection, _, progress = scenario.compose(mode="hybrid")
    html = module.MODE.build_human_page(selection, progress, scenario.rows, scenario.manifest)
    print(json.dumps({"html": html, "required": selection["requiredHumanReviewCount"], "rows": len(scenario.rows)}))
finally:
    scenario.tearDown()
`, path.join(directory, "test_review_modes.py")], { encoding: "utf8", maxBuffer: 2_000_000 });
      assert.equal(generated.status, 0, generated.stderr);
      const mode = JSON.parse(generated.stdout);
      const originalHtml = scenario.html;
      const originalStorage = await page.evaluate(() => ({ ...localStorage }));
      scenario.html = mode.html;
      const modePage = await context.newPage();
      try {
        await modePage.goto(url);
        assert.match(await modePage.locator("#progress-text").textContent(), new RegExp("0 / " + mode.rows));
        const required = modePage.locator("section").filter({ has: modePage.getByRole("button", { name: "다음 필수 사람 검토" }) });
        assert.match(await required.textContent(), new RegExp("미완료 " + mode.required));
        await modePage.getByRole("button", { name: "다음 필수 사람 검토" }).click();
        assert.match(await required.textContent(), /현재 공고: 필수 검토 대상/);
        await modePage.locator("#reviewer").fill("합성 사람 검토자");
        await modePage.locator('[data-decision="irrelevant"]').click();
        assert.match(await required.textContent(), new RegExp("미완료 " + (mode.required - 1)));
        const saved = await download(modePage);
        assert.equal(saved.data.judgments.filter((item) => item.decision !== "").length, 1);
        assert.equal(saved.data.judgments.filter((item) => item.provenance?.kind === "ai_consensus").length, 0);
        const storageAfter = await modePage.evaluate(() => ({ ...localStorage }));
        for (const [key, value] of Object.entries(originalStorage)) assert.equal(storageAfter[key], value);
        assert.equal(Object.keys(storageAfter).some((key) => key.includes(":selection:")), true);
      } finally {
        scenario.html = originalHtml;
        await modePage.close();
      }
    });
    assert.deepEqual(externalRequests, [], "The review page attempted non-local requests");
    assert.deepEqual(pageErrors, [], "Unexpected browser JavaScript errors");
    t.diagnostic("No outbound page requests; all contexts use temporary, isolated browser storage.");
  } finally {
    if (browser) await browser.close();
    if (server) await new Promise((resolve) => server.close(resolve));
    await fs.rm(temporary, { recursive: true, force: true });
  }
});
